python/ray/tune/cloud.py

```diff
@@ -0,0 +1,321 @@
+                the checkpoint. Defaults to ``False``.
+        """
+        temp_dirs = set()
+        # Per default, save cloud checkpoint
+        if not path:
+            if self.cloud_path and self.local_path:
+                path = self.local_path
+            elif not self.cloud_path:
+                raise RuntimeError(
+                    "Cannot save trial checkpoint: No cloud path "
+                    "found. If the checkpoint is already on the node, "
+                    "you can pass a `path` argument to save it at another "
+                    "location.")
+            else:
+                # No self.local_path
+                raise RuntimeError(
+                    "Cannot save trial checkpoint: No target path "
+                    "specified and no default local directory available. "
+                    "Please pass a `path` argument to `save()`.")
+        elif not self.local_path and not self.cloud_path:
+            raise RuntimeError(
+                f"Cannot save trial checkpoint to cloud target "
+                f"`{path}`: No existing local or cloud path was "
+                f"found. This indicates an error when loading "
+                f"the checkpoints. Please report this issue.")
+
+        if is_cloud_target(path):
+            # Storing on cloud
+            if not self.local_path:
+                # No local copy, yet. Download to temp dir
+                local_path = tempfile.mkdtemp(prefix="tune_checkpoint_")
+                temp_dirs.add(local_path)
+            else:
+                local_path = self.local_path
+
+            if self.cloud_path:
+                # Do not update local path as it might be a temp file
+                local_path = self.download(
+                    local_path=local_path,
+                    overwrite=force_download,
+                    update_local_path=False)
+
+            # We should now have a checkpoint available locally
+            if not os.path.exists(local_path) or len(
+                    os.listdir(local_path)) == 0:
+                raise RuntimeError(
+                    f"No checkpoint found in directory `{local_path}` after "
+                    f"download - maybe the bucket is empty or downloading "
+                    f"failed?")
+
+            # Only update cloud path if it wasn't set before
+            cloud_path = self.upload(
+                cloud_path=path,
+                local_path=local_path,
+                clean_before=True,
+                update_cloud_path=not self.cloud_path)
+
+            # Clean up temporary directories
+            for temp_dir in temp_dirs:
+                shutil.rmtree(temp_dir)
+
+            return cloud_path
+
+        local_path_exists = self.local_path and os.path.exists(
+            self.local_path) and len(os.listdir(self.local_path)) > 0
+
+        # Else: path is a local target
+        if self.local_path and local_path_exists and not force_download:
```

Comment:
there are a lot of if elses here, handling local vs cloud, etc cases. but if we make local storage also a "provider", you don't need to have any special if handling here.
supposedly if source is a local path, we can just use the local provider module to handle it, same as any other cloud provider. and this part of the logic will be much cleaner.

Reply:
A difference here is that we explicitly distinguish between a local and a cloud location in the object and would likely want to do so in terms of API.
However, I agree, once we have a Provider API, we can just do `download()` and then `copytree` the local directory.

Reply:
My proposal here would be, as long as everyone is fine with the API, to go with the current implementation to enable this case, and go to a more general storage solution later (presumably whilst refactoring the syncing logic generally)

Reply:
happy to discuss why we need to differentiate and if we can hide the difference behind a common api.
but that's definitely for a separate change.
thanks.